apps/consulting/posts/review-functorch-contribution.md

```diff
@@ -0,0 +1,323 @@
+incorporated into the [jacrev PR](https://github.com/pytorch/pytorch/pull/89376)
+and the [vmap PR](https://github.com/pytorch/pytorch/pull/91157).
+
+## Support for `linearize` transform
+
+The "jvp" transform is designed to calculate both `f(x)` and the Jacobian-vector
+product. Consequently, even when one intends to compute the Jacobian-vector
+product for fixed inputs, the `jvp` transform still redundantly evaluates `f(x)`.
+To address such scenarios, the `linearize` transform comes into play. This
+transform proves valuable when multiple `jvp` computations are needed for
+constant inputs.
+
+Note that, in order to implement this efficiently, `linearize` stores some
+intermediate computations, which can result in higher memory requirements compared
+to directly applying `jvp`. The `linearize` transform was implemented in this
+[PR](https://github.com/pytorch/pytorch/pull/94173).
+
+## Support of `torch.func` transforms within `torch.compile`
+
+PyTorch 2.0 introduced a JIT compiler under `torch.compile`, similar to `jax.jit`.
+This opened up the possibility of compiling the existing transforms to enhance
+their performance. To understand how these transforms can be compiled, it is
+essential to discuss the workings of the three layers within the compilation
+stack, namely `dynamo`, `aot_autograd`, and `inductor`.
+
+The `dynamo` and `aot_autograd` layers primarily focus on capturing the
+computation graph and converting the captured operations into more basic operations.
+This captured graph is then passed to `inductor`, the compiler. `inductor` then applies
+various optimization passes before generating specialized code.
+
+To gain insight into the different stages of this stack,
+let us compile a simple program in debug mode.
+
+```python
+# Run this file with `TORCH_COMPILE_DEBUG=1`
+
+import torch
+
+def fn(x):
+    return torch.sin(x) + torch.square(x)
+
+torch.compile(fn)(torch.randn(4, 4))
+```
+
+**dynamo**: The primary responsibility of dynamo is to trace the Python program
+and convert it into the FX graph format. The FX graph generated by `dynamo`
+represents PyTorch operations from the public API, such as `torch.sin`.
+Below, you can observe the graph captured by `dynamo`.
+
+```python
+class GraphModule(torch.nn.Module):
+    def forward(self, L_x_ : torch.Tensor):
+        l_x_ = L_x_
+
+        # File: test/test_scratch.py:334, code: return torch.sin(x) + torch.square(x)
+        sin = torch.sin(l_x_)
+        square = torch.square(l_x_);  l_x_ = None
+        add = sin + square;  sin = square = None
+        return (add,)
+```
+
+**aot_autograd**: `aot_autograd` retraces all PyTorch operations to
+produce a lower-level FX graph using `aten` functions (from the private API).
+Additionally, `aot_autograd` decomposes composite operations into primitive
+operations. For instance, a composite operation like `torch.square` is traced
+down to `torch.pow(x, 2)`.
```

Comment:
If we're being pedantic, this doesn't happen at the torch IR level. it goes `torch.square` -> `aten.square` -> `aten.pow`
```suggestion
down to `aten.pow(x, 2)`.
```